### PR TITLE
sync: update from Slime and upgrade to latest vLLM nightly

### PR DESCRIPTION
**Pull request by aoshen02**
## Summary

- synchronize Vime with Slime commits `680824dd..41014d1f` (through `#2276`)
- upgrade the CUDA image and integration baseline to the latest vLLM nightly
- add model/weight version propagation across generation and weight updates
- upgrade the trainer-side vLLM IPC weight-update API to one worker-rank-indexed
  `list[dict | None]` payload, matching vLLM PR #52497
- add native sparse disk-delta production/loading, external-engine fanout, full-disk fallback, docs, and tests
- port GLM5 alignment/indexer integration to the current vLLM nightly API surface
- keep all native vLLM integration, including the sparse disk-delta receiver, in the single
  `docker/patch/latest/vllm.patch`

## Pinned references

- Vime base: `8144096e3f4fb0fb670c37b8f2d84015f7e92320`
- Slime cutoff: `680824dd5e01a2e83750bf87fc366ec6fa98766c`
- Slime tip: `41014d1f29e201137fdffce737bb8bac65bc5219`
- final Vime commit: `c801fa89bd164c65df6158e036b930ffb70fb498`
- vLLM nightly: `0.27.2rc1.dev150+g311b3513a` at `311b3513af33bc29b4acb2fde2e9313e5e9966a0`
- vLLM PR #50902 merge: `21ea5b4fa1062a379dd7e6795497ad6becd5a856` (ancestor of the image commit)
- vLLM PR #49577 merge: `50ba4bc6b2cacd70c4711a1904dd7ad9740a578b` (ancestor of the image commit)
- vLLM delta loader reference: vllm-project/vllm#50723 at `fd07acd5b596c11f949fa71b5f0ee926b9e6bf17`
- VERL integration reference: verl-project/verl#7227 at `e6ca5e7df49c6db564969f8f9aee39ef3e764311`

## Delta/version behavior

- the trainer captures a baseline HF tensor snapshot, emits sparse changed-byte artifacts, and uses Slime's monotonically increasing trainer-side weight version
- vLLM reconstructs a canonical host-local checkpoint from sparse artifacts, reloads it through
  its ordinary checkpoint loader, and reports the applied weight version in generation responses
- failed fanout does not finish the engine-side weight-update session or resume generation
- disk transport retains the full-checkpoint path as a non-delta fallback
- colocated IPC sends one update RPC per chunk; each vLLM worker selects its rank-local payload,
  and `None` skips a worker without the removed UUID-grouping protocol

## Validation

- full pre-commit: all hooks passed
- every merge-order Slime PR in `680824dd..41014d1f` was audited: 40 PRs, no stale
  deletion, no upstream-touched path frozen at the old mirror, and no unexplained missing path
- the single `docker/patch/latest/vllm.patch` applies to the pinned nightly vLLM tree; the former
  separate delta patch is removed, and `docker/README.md` is bitwise equal to the clean Vime base
- the `#2267/#2276` argument/conversion regression set passed `38/38`; `#2274` is retained for
  the authoritative AMD CI compile/runtime gate
- 33 synchronized upstream CPU/unit test files passed in isolated Python processes, matching CI
- final nightly vLLM patch tests: rank-local worker/client `11 passed`; delta receiver `4 passed`
- final focused current-source Vime updater/alignment tests: `274 passed`
- the single `docker/patch/latest/vllm.patch` applies to the pinned nightly vLLM tree; the fully
  patched tree contains no old `empty_gpu_uuids` protocol
- H200 external-PD Megatron/vLLM sparse-delta training: SLURM job `630`, `COMPLETED`, `ExitCode=0:0`
- H200 colocated IPC/fanout training: SLURM job `641`, two rollout/training steps passed
- H200 full-disk release training: SLURM job `635`, two release/recreate steps passed
- H200 Qwen3.6-35B-A3B PD/DP4/EP/DeepEP regression at commit `17ecedbac`: complete test
  exited `0`; both initial and post-train IPC updates returned `200` for every DP replica, with
  no GPU-UUID mismatch (`agent_run/results/qwen36-dp4-ipc-fix-20260816/test.log`)
- H200 GLM4.7-30B-A3B PD/DP4/EP regression at commit `17ecedbac`: complete test exited `0`;
  the initial and both post-train IPC updates returned `200` for every DP replica, with no
  GPU-UUID mismatch (`agent_run/results/glm47-dp4-ipc-fix-20260816/test.log`)
- H200 Qwen3-4B PPO regression at commit `17ecedbac`: complete test exited `0`; two actor/critic
  steps and all four colocated TP2 IPC update paths completed without NCCL/CUDA or UUID-routing
  failures (`agent_run/results/qwen3-4b-ppo-current-20260816/test.log`)
- H200 full-disk update regression at commit `17ecedbac`: complete test exited `0`; both
  Hugging Face checkpoint versions reloaded through vLLM and advanced the published weight
  version (`agent_run/results/full-disk-current-20260816/test.log`)
- H200 Qwen3-4B PPO disaggregation regression at commit `17ecedbac`: complete test exited `0`;
  two actor/critic steps and all external TP2 NCCL update rounds succeeded
  (`agent_run/results/qwen3-4b-ppo-disaggregate-current-20260816/test.log`)
- the GPU run captured 398 baseline tensors, trained through step 2, and applied sparse delta versions through v4
- train/rollout absolute log-prob diff: `0.00838`, `0.01034`, `0.00768` for steps 0-2 (threshold `0.1`)
- non-empty delta density/wire size: v2 `1.61% / 389,240,350 B`, v3 `1.31% / 315,475,560 B`, v4 `1.23% / ~0.30 GB`
- controlled Slime/SGLang delta-disk A/B: steady `7.9 s → 7.0 s` (`1.13×`), payload
  `~8.05 GB → 0.13-0.17 GB`; first real delta averaged `21.8 s`
- controlled Vime/vLLM delta-disk A/B: delta steady `8.0-8.3 s` at `0.14-0.15 GB`, full
  `7.4-7.8 s`; this local 4B case validates the path but is not claimed as a latency win
- matched two-node Qwen3-30B-A3B A/B (`h200-0` trainer 8 GPU, `h200-1` rollout 8 GPU):
  Vime/Slime NCCL medians are `17.221 s` / `17.881 s`; nonzero delta-disk updates are
  `39.414 s` / `47.627 s` with `0.50 GB` / `0.56 GB` payloads. The delta bottleneck is
  trainer encode plus receiver apply, so this 30B run validates payload reduction and semantic
  parity but does not claim a latency win. Full phase table and raw logs:
  `agent_run/results/weight30b-two-node-noep-ab-20260816T124448Z/summary.md`

## Final nightly validation

- every non-mechanical hunk under `vime_plugins/`, `vime/utils/`, and
  `vime/backends/megatron_utils/` was re-audited against the mechanical Slime mirror; the final
  31-file ledger contains only required vLLM translations or previously approved Vime divergences
- every one of the 78 documentation-tree differences was audited; both English and Chinese Sphinx
  builds pass without candidate-only warning classes
- GLM5.2 CUDA graph sizing is translated explicitly: shared batch `8 × (1 + 5) = 48`, decode
  group `12 × (1 + 5) = 72`
- parent nightly image focused H200 suite passed `104/104`
- final immutable image NCCL adapter unit tests passed `11/11`; exact 8-GPU disaggregated PPO passed with `train_rollout_logprob_abs_diff=0.0151497051`
- final immutable image local H200 precision matrix passed all 12 Ray subjobs across both loss modes and DP/TP/PP/CP layouts
- new H200 E2E coverage passed: non-colocated NCCL Moonlight-16B-A3B, Qwen3.5 top-p/CP2, and
  Qwen3.5-VL train/rollout

Authoritative image: `vllm/vime:latest`, remote manifest digest
`sha256:16a605fca3e29311b7ffab76a71a4635ebc78479693ab30136c868690d0b219a`.
Runtime: Vime `c801fa89bd164c65df6158e036b930ffb70fb498`, vLLM
`0.27.2rc1.dev150+g311b3513a` (`311b3513af33bc29b4acb2fde2e9313e5e9966a0`), H200.

Authoritative remote CI for current commit `c801fa89bd164c65df6158e036b930ffb70fb498`:
[Buildkite #935](https://buildkite.com/vllm/vime-ci/builds/935), running all six GPU suites
(`short`, `vllm-config`, `megatron`, `vime-customized`, `precision`, and `ckpt`).

## 100-step Qwen3-30B-A3B convergence regression

- W&B: [run `gzzec11f`](https://wandb.ai/aoshen524-purdue-university/vime-ab-30b-h200/runs/gzzec11f)
- workload: Issue #79 recipe on one 8×H200 node; colocated Megatron/vLLM, 32 prompts × 16
  samples, global batch 512, 20,480 response cap, 32,768 context cap, evaluation every 20
  updates, and full rank-local IPC weight synchronization
- completion: all iterations `0..99`, every scheduled evaluation, the iteration-99 checkpoint,
  Ray job, and container completed successfully

| Completed updates | AIME pass@1 | AIME pass@32 |
|---:|---:|---:|
| 0 | 0.9375% | 16.6667% |
| 20 | 2.8125% | 36.6667% |
| 40 | 13.3333% | 43.3333% |
| 60 | 17.1875% | 40.0000% |
| 80 | 17.7083% | 46.6667% |
| 100 | **20.3125%** | **50.0000%** |

- AIME reward improved from `-0.98125` to `-0.59375`; final pass@2/4/8/16/32 was
  `0.255645 / 0.310718 / 0.367705 / 0.432708 / 0.500000`
- final AIME truncated ratio was `8.6458%`, median response length was `938.5`, and the final
  train/rollout log-prob absolute difference was `0.00698687` (threshold `0.1`)
- IPC update latency across all 100 updates was min/mean/max
  `3.314952 / 4.001741 / 5.020601 s`; ten-step means stayed in `3.818–4.204 s`, so there was
  no progressive synchronization slowdown
- versus the previous matched 100-step vLLM 0.24 run, final pass@1 was
  `21.2500% → 20.3125%` and pass@32 stayed `50.0000%`; mean weight-update time improved
  `25.339 → 4.002 s` (`6.33×`) and median step time improved `183.020 → 151.962 s`
- this is not a pure IPC microbenchmark because vLLM, PyTorch, Vime, and rollout expert-parallel
  execution also changed; it demonstrates preserved convergence plus stable end-to-end IPC behavior
- provenance: source manifest `92f188c2201487836c4903eecf5804e48540b445a7158d3317131e0d44057fb6`,
  runtime vLLM `0.27.2rc1.dev150+g311b3513a`; the copied runtime code matches rebased Vime
  `b33a817d` except documentation. The later `c801fa89` change is confined to the non-colocated
  packed-NCCL path and is validated separately by the exact 8-GPU NCCL E2E above.

The final GPU harness copies source into the container rather than bind-mounting it. `NCCL_NVLS_ENABLE=0` is set for this host because its Fabric Manager rejects NVLS initialization; ordinary NCCL paths remain exercised.

For the PyTorch `2.13.0+cu130` image, Vime keeps native process groups alive while offloading model tensors. Python `ProcessGroup` wrappers reproducibly segfault in `PyWorkHolder.wait` on this runtime; retaining native groups avoids that crash with a small communicator-memory tradeoff.

image

### CODE REVIEW

File: vime/backends/megatron_utils/update_weight/common.py [hunk 1]
```diff
@@ -214,24 +230,115 @@ def _named_params_and_buffers_global(
...
+        with socket.socket() as sock:
+            sock.bind(("", 0))
+            rendezvous[0] = (ray._private.services.get_node_ip_address(), sock.getsockname()[1])
+    dist.broadcast_object_list(rendezvous, src=0, group=get_gloo_group())
+    master_address, master_port = rendezvous[0]
+    return WeightTransferTrainerFactory.trainer_init(
+        NCCLTrainerInitInfo(
+            master_address=master_address,
+            master_port=master_port,
+            world_size=sum(engine_gpu_counts) + 1,
```
Comment (CalvinXKY) [inline vime/backends/megatron_utils/update_weight/common.py:338]: `create_nccl_trainer()` now initializes a trainer on every Megatron rank (`rank=dist.get_rank()`), but `world_size` is still `sum(engine_gpu_counts) + 1` and engine `rank_offset` still starts from 1 in `VimeRayWeightSyncClient.init_weight_transfer_engine()`.

This creates rank collisions as soon as trainer world size > 1: trainer ranks 1..N overlap with engine ranks 1..N. 

For example, with 2 trainer ranks and one 4-GPU engine, the communicator layout becomes trainer={0,1}, engine={1,2,3,4}, while `world_size=5`. That is inconsistent and can break distributed NCCL weight transfer.

Either keep a single trainer rank, or expand the communicator layout so trainer ranks and engine ranks occupy disjoint ranges?

Comment (aoshen02) [reply vime/backends/megatron_utils/update_weight/common.py:338]: Nice catch. All trainer ranks construct a trainer-side engine, but NCCLTrainerInitInfo.rank is only used to select the sender. Non-zero trainer ranks return before opening the NCCL endpoint or initializing the inference workers, so only trainer rank 0 joins the transfer communicator. This multi-rank path has also been validated in CI.

**Review by CalvinXKY — APPROVED** [review]
